### PR TITLE
Grow the built-in function library (RFC 0007 / R-33)

### PR DESCRIPTION
**Pull request by Evgenus**
## Summary
- Add 30 new `call` functions (string/case, slice, epoch dates, collections/numerics, bool, base64, uuid5, regex), a `split` rule (strings + arrays, `NO_CONTENT` passthrough), and a total binary `in` membership operator — all additive, with failures raised as `TransformationError`.
- Spec, corpus examples, error-path tests, and ROADMAP R-33 marked done under `[Unreleased]`; no version bump (0.2.0 held with RFC 0008 language-reference export).
- Closes the measured authoring capability gaps (uppercase, prefix strip, epoch→ISO) and collapses common reduce idioms (`length`/`flatten`/`sum`).

## Test plan
- [x] `uv run pytest .` (378 passed; `functions.py` 100% coverage)
- [x] `uv run python -m transon.docs` — no TBD
- [x] `python scripts/check_roadmap.py` — consistent
- [ ] CI green on this PR
- [ ] Spot-check metadata: `get_editor_metadata()` lists new functions/`in`/`split` with example refs

Made with [Cursor](https://cursor.com)

## Summary by CodeRabbit

* **New Features**
  * Expanded the built-in function library with new string, numeric, collection, date/epoch, encoding (Base64), UUIDv5, and regex utilities.
  * Added the `split` transformation rule for strings and arrays, including `NO_CONTENT` passthrough behavior.
  * Added a total `in` membership operator for arrays, strings, and object keys that never errors.

* **Documentation**
  * Updated the specification, roadmap, and release notes to reflect the new rule/operator/function set and contracts.

* **Tests**
  * Added comprehensive documentation and edge-case coverage for the new functions, `split`, and `in`, including error-path validation.

### CODE REVIEW

File: docs/proposals/0007-builtin-function-library.md [hunk 1]
```diff
@@ -1,9 +1,9 @@
 # RFC 0007 — Grow the built-in function library (string / numeric / collection helpers)
 
-- **Status:** Proposed — `needs-decision`
+- **Status:** Implemented — pending 0.2.0 packaging (held with RFC 0008)
 - **Created:** 2026-07-16
-- **Roadmap:** R-33 (`needs-decision`) — targets a **0.2.0** minor release
-- **Type:** Additive engine capability — every addition is a newly-named function/operator; no behavior change to existing templates
+- **Roadmap:** R-33 (`done`) — targets a **0.2.0** minor release (with RFC 0008)
+- **Type:** Additive engine capability — every addition is a newly-named function/operator/rule; no behavior change to existing templates
```
[the remark this answers is not in the document]
Comment (Evgenus) [reply docs/proposals/0007-builtin-function-library.md]: Fixed — status is now **Accepted** (code landed; packaging deferred to 0.2.0 with RFC 0008), with an explicit note that `Implemented` is reserved for a shipped release. Index row updated to match.

File: transon/functions.py [hunk 2]
```diff
@@ -1,10 +1,50 @@
...
 from transon import (
     Transformer,
     DefinitionError,
+    TransformationError,
 )
 from transon.transformers import format_error_message
 
 
+_ISO_EPOCH_FMT = '%Y-%m-%dT%H:%M:%SZ'
+_EPOCH_FMT_ALLOWED = frozenset('YmdHMSj z%')
```
[the remark this answers is not in the document]
Comment (Evgenus) [reply transon/functions.py]: Fixed — removed the stray space from `_EPOCH_FMT_ALLOWED` (`frozenset('YmdHMSjz%')`). Added a regression test that `%Y% %m` is rejected.

File: transon/functions.py [hunk 3]
```diff
@@ -30,6 +70,308 @@ def _json_type(value):
...
+        _validate_epoch_fmt(fmt)
+    try:
+        dt = datetime.strptime(s, fmt)
+    except ValueError as exc:
+        _error(f'`to_epoch` failed to parse {s!r} with format {fmt!r}: {exc}')
+    if dt.tzinfo is None:
+        dt = dt.replace(tzinfo=timezone.utc)
+    else:
+        dt = dt.astimezone(timezone.utc)
+    return int(dt.timestamp())
```
[the remark this answers is not in the document]
Comment (Evgenus) [reply transon/functions.py:185]: Fixed the actionable sibling — `b64encode` now wraps `UnicodeEncodeError` (e.g. lone surrogates) as `TransformationError`. `_to_epoch` left unchanged: aware UTC timestamps are computed without platform `mktime`, so no extra OverflowError guard is needed there.

File: transon/functions.py [hunk 4]
```diff
@@ -30,6 +70,308 @@ def _json_type(value):
...
+    pattern = _require_str(pattern, 'regex_match')
+    try:
+        match = re.search(pattern, s)
+    except re.error as exc:
+        _error(f'`regex_match` invalid pattern: {exc}')
+    if match is None:
+        return None
+    if match.lastindex:
+        return [g for g in match.groups()]
+    return [match.group(0)]
```
[the remark this answers is not in the document]
Comment (Evgenus) [reply transon/functions.py:365]: Fixed — decision now uses `match.re.groups` (pattern capture count) instead of `lastindex`, so `regex_match('b', r'(a)?b')` returns `[null]`. Regression test added.

File: transon/functions.py [hunk 5]
```diff
@@ -49,3 +391,158 @@ def _json_type(value):
...
+        '(no random `uuid4`).',
+)(_uuid5)
+Transformer.register_function(
+    'regex_match', input_type='string', output_type='array',
+    doc='`values: [s, pattern]` — Python `re.search`. On match returns capture groups '
+        'as an array (unmatched optionals are `null`); with no groups returns '
+        '`[full match]`. On no match returns `null`. Use `bool(...)` in conditions. '
+        'Invalid pattern raises `TransformationError`. Dialect is Python `re`; ReDoS '
+        'is a host responsibility.',
+)(_regex_match)
```
[the remark this answers is not in the document]
Comment (Evgenus) [reply transon/functions.py]: Fixed — `output_type` is now `array, null`, and the embedded `call` name table in `rules.py` matches.

File: transon/rules.py [hunk 6]
```diff
@@ -487,6 +487,81 @@ def rule_join(t: Transformer, template, context: Context):
...
+            )
+        return value.split(sep)
+    if isinstance(value, list):
+        if isinstance(sep, list):
+            if not sep:
+                t.transformation_error(
+                    '`sep` must be a non-empty array when used as a subsequence'
+                )
+            return _split_array_subsequence(value, sep)
+        return _split_array_element(value, sep)
```
[the remark this answers is not in the document]
Comment (Evgenus) [reply transon/rules.py:531]: Fixed — `sep is NO_CONTENT` now raises `TransformationError` before string/array mode selection. Regression test uses `get` of an undefined name as the sep template.

**Comment by Evgenus** [comment]
Addressed all six CodeRabbit actionable findings (plus the two quick nitpicks: template-path assert on one error test, and an `op:in` substring corpus case) in f62b80e.
